pkg/cache/output_predictor.go

```diff
@@ -0,0 +1,265 @@
+	atomic.AddInt32(&hist.size, -hist.Tail().getSkipped())
+}
+
+func NewSimmpleOutputPredictor(maxInputTokens, maxOutputTokens int, window time.Duration) *SimmpleOutputPredictor {
+	// We allocate 1 more history slot to make summary update on rotating lock free
+	extraSlot := 1
+	if window%MovingInterval > 0 {
+		extraSlot++
+	}
+	inputBuckets := int(math.Ceil(math.Log2(float64(maxInputTokens + 1))))
+	outputBuckets := int(math.Ceil(math.Log2(float64(maxOutputTokens + 1))))
+	predictor := &SimmpleOutputPredictor{
+		history: rotatingHistory{
+			window:        make([]outputDistribution, int(window/MovingInterval)+extraSlot),
+			headTimestamp: time.Now(),
+		},
+		inputs:        make(outputDistribution, inputBuckets*outputBuckets),
+		inputsSums:    make([]int32, inputBuckets),
+		inputBuckets:  inputBuckets,
+		outputBuckets: outputBuckets,
+		rand:          rand.Int31n,
+	}
+	for i := 0; i < len(predictor.history.window); i++ {
+		predictor.history.window[i] = make(outputDistribution, inputBuckets*outputBuckets+1)
+	}
+	return predictor
+}
+
+func (p *SimmpleOutputPredictor) AddTraceWithTimestamp(inputTokens, outputTokens int, cnt int32, ts time.Time) {
+	p.tryRotate(ts)
+
+	inputBucket := p.token2bucket(inputTokens, p.inputBuckets)
+	idx := p.bucket2idx(inputBucket, p.token2bucket(outputTokens, p.outputBuckets))
+
+	// Avoid operations during rotating
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	// Add summary first and history next to avoid possible negative summary on rotating.
+	atomic.AddInt32(&p.inputs[idx], cnt)
+	atomic.AddInt32(&p.inputsSums[inputBucket], cnt)
+	atomic.AddInt32(&p.history.window[p.history.head][idx], cnt)
+}
+
+func (p *SimmpleOutputPredictor) AddTrace(inputTokens, outputTokens int, cnt int32) {
+	p.AddTraceWithTimestamp(inputTokens, outputTokens, cnt, time.Now())
+}
+
+func (p *SimmpleOutputPredictor) Predict(inputTokens int) int {
+	inputBucket := p.token2bucket(inputTokens, p.inputBuckets)
+	randRange := atomic.LoadInt32(&p.inputsSums[inputBucket])
+	if randRange == int32(0) {
+		return p.coldPredict(inputTokens) // No history, return input tokens
+	}
+	// Do weighted random
+	cursor := p.rand(randRange)
+	accumulation := int32(0)
+	scanRange := (inputBucket + 1) * p.outputBuckets
+	for i := scanRange - p.outputBuckets; i < scanRange; i++ {
+		accumulation += atomic.LoadInt32(&p.inputs[i])
+		if cursor < accumulation {
+			return int(math.Pow(2, float64(i-scanRange+p.outputBuckets)))
+		}
+	}
+	return int(math.Pow(2, float64(p.outputBuckets-1)))
+}
+
+func (p *SimmpleOutputPredictor) coldPredict(inputTokens int) int {
+	switch DefaultColdPrediction {
+	case RandomColdPredition:
+		return rand.Intn(MaxOutputLen) + 1
+	case InputColdPrediction:
+		return inputTokens
+	case PessimiticColdPrediction:
+		return MaxOutputLen
+	default:
+		// Default to optimistic
+		return 1
+	}
+}
+
+func (p *SimmpleOutputPredictor) bucket2idx(inputBucket, outputBucket int) int {
+	return inputBucket*p.inputBuckets + outputBucket
```

Comment:
is this used for mapping the 2D dimension -> single dimension? in that case, should it be `inputBucket*p.outputBuckets`